compiler/resolution/functionResolution.cpp

```diff
@@ -10049,17 +10049,19 @@ static void removeUnusedTypes() {
   }
 }
 
-static void removeUnusedGlobals()
-{
-  forv_Vec(DefExpr, def, gDefExprs)
-  {
-    // Remove unused global variables
-    if (toVarSymbol(def->sym))
-      if (toModuleSymbol(def->parentSymbol))
-        if (def->sym->type == dtUnknown ||
-            (def->sym->type->symbol->hasFlag(FLAG_GENERIC) &&
-             def->sym->hasFlag(FLAG_TYPE_VARIABLE)))
-          def->remove();
+// Remove module level variables if they are not defined or used
+// With the exception of variables that are defined in the rootModule
+static void removeUnusedModuleVariables() {
+  forv_Vec(DefExpr, def, gDefExprs) {
+    if (VarSymbol* var = toVarSymbol(def->sym)) {
+      if (ModuleSymbol* module = toModuleSymbol(def->parentSymbol)) {
+        if (var->isDefined() == false && var->isUsed() == false) {
```

Comment:
I think this is a reasonable way to write it, but knowing what isDefined and isUsed actually do - you're actually just checking if the Symbol is referred to by any SymExprs that are in the tree. That's even simpler than isDefined/isUsed (which check for SymExprs in specific contexts - uses or defs). So it would seem that we could have a Symbol method that just returns true if for_SymbolSymExprs iterates over anything at all. The main problem I see there is that it's hard to come up with a good name for it.

Either way this is a minor issue & I'm glad that this code is simplified and more correct.